platform_api/orchestrator/poller_service.py

```diff
@@ -139,67 +140,91 @@ def _filter_our_for_quota(
         now = self._current_datetime_factory()
 
-        # Always start/update not scheduled jobs
-        jobs_to_update.extend(job for job in unfinished if not job.scheduler_enabled)
+        unfinished = await self._enforce_running_job_quota(unfinished)
 
-        scheduled = [job for job in unfinished if job.scheduler_enabled]
-
-        not_materialized = sorted(
-            (job for job in scheduled if not job.materialized),
-            key=lambda job: job.status_history.current.transition_time,
-        )
-        materialized_not_running = [
+        not_materialized = [job for job in unfinished if not job.materialized]
+        pending = [
             job
-            for job in scheduled
+            for job in unfinished
             if job.status != JobStatus.RUNNING and job.materialized
         ]
-        running = [job for job in scheduled if job.status == JobStatus.RUNNING]
+        not_scheduled_running = [
+            job
+            for job in unfinished
+            if not job.scheduler_enabled and job.status == JobStatus.RUNNING
+        ]
+        scheduled_running = [
+            job
+            for job in unfinished
+            if job.scheduler_enabled and job.status == JobStatus.RUNNING
+        ]
 
-        waiting_exists = any(
-            now - job.status_history.current.transition_time
+        # Process jobs waiting for resources
+        waiting = [
+            job
+            for job in pending
+            if now - job.status_history.current.transition_time
             >= self._config.is_waiting_min_time
-            for job in materialized_not_running
-        )
+        ]
+        if waiting:
+            # Handle one priority per poller iteration to avoid
+            # replace of higher priority jobs with lower priority jobs.
+            jobs_waiting_priority = max(job.priority for job in waiting)
+            jobs_waiting.extend(
+                job for job in waiting if job.priority == jobs_waiting_priority
+            )
 
-        # not_materialized processing
+        # Always start/update pending and not scheduled jobs
+        jobs_to_update.extend(pending)
+        jobs_to_update.extend(not_scheduled_running)
 
+        # Process suspended jobs
+        has_suspended = False
         for job in not_materialized:
-            suspended_at = job.status_history.current.transition_time
-            if (
-                not waiting_exists
-                or now - suspended_at >= self._config.max_suspended_time
-            ):
-                jobs_to_update.append(job)
-
-        # materialized_not_running processing
-
-        jobs_to_update.extend(materialized_not_running)
-
-        # running processing
-
-        for job in running:
+            if job.is_suspended:
+                has_suspended = True
+                suspended_at = job.status_history.current.transition_time
+                if now - suspended_at >= self._config.max_suspended_time:
```

Comment:
Hmm, looks like here you change logic a bit.

Previously, we resumed suspended jobs only if there is no waiting jobs (so there *might* be enough resources to fit additional job). It is sub-optimal, as it can happen that a waiting job requires GPU, and a suspended job needs only CPU and the cluster has enough resources. To partially overcome this issue, we resume jobs if they have waited for too long.

So SUSPENDED state here is a "long" state - a job can be in this state for hours. 

But you've changed `max_suspended_time` default to 1:30 minutes, and also do not schedule new jobs while there is at least a single SUSPENDED job. So I guess you want SUSPENDED state to be "short", or it will block new jobs.

My guess is that your idea is to use k8s internal queue instead of our queue of suspended jobs as much as possible. I think it will not work properly for the following case:

You have a cluster with 0.2 CPU, and 4 jobs - 3x 0.1cpu, and 1x0.2cpu, all with the same priority. If we materialize the suspended jobs almost instantly, then 0.2cpu job will never run.